src/main.cpp

```diff
@@ -1606,6 +1606,13 @@ bool AcceptToMemoryPool(CTxMemPool& pool, CValidationState &state, const CTransa
     return true;
 }
 
+bool GetSpentIndex(CSpentIndexKey &key, CSpentIndexValue &value)
+{
+    if (!fSpentIndex)
+        return false;
+    return pblocktree->ReadSpentIndex(key, value);
```

Comment:
This requires holding the `cs_main` lock. `GetSpentIndex` is only called from `TxToJSON`, and it just so happens that we do hold the lock at that time. But I'm somewhat surprised that the upstream patchset includes a change (which this PR doesn't pull in) that means the lock would not be held here. Additionally, IIRC there's a recent Bitcoin Core PR that also drops the lock when encoding JSON (for performance reasons). So it's possible in future that a caller would not be holding the necessary lock.

We should either assert that `cs_main` is locked here, or explicitly take it. Probably the former, to force callers to set up their locking correctly.

Reply:
The upstream Bitcoin Core PRs I was thinking of are the ones that have been backported in #4020, and are for `getblock` rather than `getrawtransaction`, but my point stands.

Reply:
This will do for now.